EIPS/eip-2315.md

```diff
@@ -0,0 +1,91 @@
+---
+eip: 2315
+title: Simple Subroutines for the EVM
+status: Draft
+type: Standards Track
+category: Core
+author: Greg Colvin
+discussions-to: https://ethereum-magicians.org/t/eip-2315-simple-subroutines-for-the-evm/3941
+created: 2019-10-17
+---
+## Abstract
+
+This proposal introduces two opcodes to support subroutines:  `JUMPSUB` and `RETSUB`.
+
+## Motivation
+
+The EVM does not provide subroutines as a primitive.  Instead, calls must be synthesized by fetching and pushing the current program counter on the data stack and jumping to the subroutine address; returns must be synthesized by getting the return address to the top of stack and jumping back to it.
+
+## Specification
+
+**_JUMPSUB_**
+
+* Jumps to the address on top of the stack.  This must be a `JUMPDEST`.
+
+**_RETSUB_**
+
+* Returns to the most recently executed `JUMPSUB` instruction.
+
+## Rationale
+
+This is the smallest possible change that provides native subroutines without breaking backwards compatibility.  It does not enforce validity like [EIP-615](https://eips.ethereum.org/EIPS/eip-615), but code that
+* pushes a constant argument before every `JUMP`, `JUMPI` and `JUMPSUB`
+* ensures that the _data stack_ is the same size at every execution of each `JUMP`, `JUMPI`, `JUMPSUB` or `JUMPDEST` operation.
+
+is valid and will meet all of the [EIP-615](https://eips.ethereum.org/EIPS/eip-615) safety conditions.  In particular, the Yellow paper defines five exceptional halting states.  
```

Comment:
That may be (re the YP), but I'm not sure that is complete. This comes to mind too:

- State-modification in a `STATICCALL` context

Also, for bullet `2` below, that should be split into two separate things: 

- More than 1024 stack items
- More than 1024 call depth

Might also want to clarify that the `Insufficient gas` is something of a catch-all, where some other errors are defined as "this error behaves as if insufficient gas was provided", e.g. if you try to use `RETURNDATACOPY` out of bounds.

Reply:
I'm yanking this.

Reply:
I know it's possible to lay out rules for using these ops to meet EIP-615's safety rules.  But it's not necessary - and not worth the trouble - to lay them out here.